### PR TITLE
Feature/event check add algo

### PR DESCRIPTION
**Pull request by AlexaBus**
Please merge check event feature with develop. It adds the following:
1. as an admin i can no longer change my role or another admin's role
2. top navbar present on all pages
3. started play eligibility algorithm and it now checks that:
 - the play does not overlap with an existing one;
 - the play is no longer then 3 hours;
 - it doesn't end after 11 PM;
4. added custom exceptions

TO DO:
1. need to catch custom exceptions in controller and show them to the user;
2. keep form completed if there are errors.

PS: it contains @mariusherta features

### CODE REVIEW

File: otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java [hunk 1]
```diff
@@ -42,6 +48,11 @@
     @Autowired
     EventService saveEvent;
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java]: Change from EventService to eventService

File: otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java [hunk 2]
```diff
@@ -42,6 +48,11 @@
...
 
+    @Autowired
+    CheckPlayEligibility eligibility;
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java]: You have EventService, you don't neet another service. Just add functionallity in addEvent method to check if all the prerequisites are satisfied in order to add a new event

File: otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java [hunk 3]
```diff
@@ -42,6 +48,11 @@
...
+
+    List<String> myExceptions = new ArrayList<>();
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java]: Check project as an inspiration StudentManagement to take a look for custom exceptions

File: otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java [hunk 4]
```diff
@@ -93,12 +109,31 @@ public String addEventSubmit(HttpServletRequest request,
         }
 
         try {
-             saveEvent.saveEvent(eventToAdd);
-        } catch ( Exception ex ) {
-            LOGGER.error(ex.getMessage());
+            if(!eligibility.checkTimeSlot(hallRepository.findByHallName(hallName).getHallName(),eventToAdd.getEventStartDate(),eventToAdd.getEventEndDate()) &&
+                    !eligibility.checkPlayLength(eventToAdd.getEventStartDate(),eventToAdd.getEventEndDate()) &&
+                    !eligibility.checkPlayEndTime(eventToAdd.getEventEndDate())) {
+                saveEvent.saveEvent(eventToAdd);
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java]: Don't put business logic in controller. In controller you just receive an event and pass to service in order to add that event, never do business logic in controller.
In controller you should have:

eventService.createEvent(event);

In method createEvent from eventService you will have to do the business logic.

All the business logic should be made in service.

File: otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java [hunk 5]
```diff
@@ -93,12 +109,31 @@ public String addEventSubmit(HttpServletRequest request,
...
+                return "redirect:/manageEvents";
+            }
+            else if(eligibility.checkTimeSlot(hallRepository.findByHallName(hallName).getHallName(),eventToAdd.getEventStartDate(),eventToAdd.getEventEndDate())) {
+                myExceptions.add(new TimeIntervalException().getMessage());
+                throw new TimeIntervalException();
+            }
+            else if (eligibility.checkPlayLength(eventToAdd.getEventStartDate(),eventToAdd.getEventEndDate())) {
+                myExceptions.add(new LengthException().getMessage());
+                throw new LengthException();
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/controller/ManageEventsController.java]: myExceptions?

File: otheatre/src/main/java/com/orange/otheatre/otheatre/exceptions/PlayEligibilityExceptions/EndTimeAfterException.java [hunk 6]
```diff
@@ -0,0 +1,9 @@
+package com.orange.otheatre.otheatre.exceptions.PlayEligibilityExceptions;
+
+public class EndTimeAfterException extends Exception {
+
+    public EndTimeAfterException() {
+        super("The play can't end after 11:00PM!");
+    }
+
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/exceptions/PlayEligibilityExceptions/EndTimeAfterException.java]: We don't need in this context a particular exception. A Conflict exception is enough.

File: otheatre/src/main/java/com/orange/otheatre/otheatre/exceptions/PlayEligibilityExceptions/LengthException.java [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package com.orange.otheatre.otheatre.exceptions.PlayEligibilityExceptions;
+
+public class LengthException extends Exception {
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/exceptions/PlayEligibilityExceptions/LengthException.java]: Same as above, we don't need this.

File: otheatre/src/main/java/com/orange/otheatre/otheatre/service/CheckPlayEligibility.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+import org.springframework.stereotype.Service;
+
+import java.time.Duration;
+import java.time.LocalDate;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.function.Predicate;
+
+@Service
+public class CheckPlayEligibility {
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/service/CheckPlayEligibility.java]: Please use EventService

File: otheatre/src/main/java/com/orange/otheatre/otheatre/service/CheckPlayEligibility.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    public Boolean checkPlayEndTime(LocalDateTime desiredEndDate) {
+        final LocalDateTime latestEndTime = LocalDate.now().atTime(23,0);
+        return desiredEndDate.isAfter(latestEndTime);
+    }
+
+    public boolean checkTimeSlot(String hallName, LocalDateTime desiredStartDate, LocalDateTime desiredEndDate) {
+        Predicate<Event> dateCheck =
+                e -> e.getEventStartDate().isBefore(desiredStartDate)&&e.getEventEndDate().isAfter(desiredStartDate) ||
```
Comment (emanueldumitru) [inline otheatre/src/main/java/com/orange/otheatre/otheatre/service/CheckPlayEligibility.java]: Instead of desiredStartDate you can user newEventStartDate

**Comment by AlexaBus** [comment]
will create a new pull request as the rebase was reverting many changes

**Pull request closed without merging**